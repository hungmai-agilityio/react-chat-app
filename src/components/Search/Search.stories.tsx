import { Meta, StoryObj } from '@storybook/react';

// FontAwesome
import { faSearch } from '@fortawesome/free-solid-svg-icons';

// Components
import Search from '@/components/Search';

const meta: Meta<typeof Search> = {
  component: Search,
  tags: ['autodocs'],
  argTypes: {
    value: {
      description: 'Input value'
    },
    icon: {
      description: 'Icon from FontAwesome'
    },
    onChange: {
      description: 'Handles change input value'
    }
  }
};

export default meta;

type Story = StoryObj<typeof Search>;

export const InputWithIcon: Story = {
  args: {
    onChange: () => {},
    value: '',
    icon: faSearch
  }
};

export const InputNotIcon: Story = {
  args: {
    onChange: () => {},
    value: ''
  }
};
