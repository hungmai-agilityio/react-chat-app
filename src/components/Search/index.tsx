import { ChangeEvent, memo } from 'react';

// FontAwesome
import { IconDefinition } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

interface SearchProps {
  value: string;
  icon?: IconDefinition;
  onChange: (event: ChangeEvent<HTMLInputElement>) => void;
}

const Search = memo(({ value, icon, onChange }: SearchProps) => {
  const handleInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    onChange(event);
  };

  return (
    <div className="relative">
      <input
        className="border border-gray-300 focus:border-primary outline-none w-full h-10 pl-4 pr-9"
        placeholder="Search"
        onChange={handleInputChange}
        value={value}
      />

      {icon && (
        <FontAwesomeIcon
          icon={icon}
          className="absolute top-3 right-3 text-gray-500 cursor-pointer"
        />
      )}
    </div>
  );
});

export default Search;
