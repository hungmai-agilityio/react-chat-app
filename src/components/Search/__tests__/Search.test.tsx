import { fireEvent, render, screen } from '@testing-library/react';

// FontAwesome
import { faSearch } from '@fortawesome/free-solid-svg-icons';

// Components
import Search from '@/components/Search';

describe('Search', () => {
  const propsDefault = {
    value: 'value',
    onChange: jest.fn()
  };

  it('render with props default', () => {
    render(<Search {...propsDefault} />);

    const element = screen.getByPlaceholderText('Search');

    expect(element).toBeInTheDocument();
  });

  it('renders with faSearch icon', () => {
    const { container } = render(<Search {...propsDefault} icon={faSearch} />);

    const iconElement = container.querySelector('svg');
    expect(iconElement).toBeInTheDocument();
    expect(iconElement).toHaveClass('fa-magnifying-glass');
  });

  it('calls onSearch when input value changes', () => {
    render(<Search {...propsDefault} />);

    const inputElement = screen.getByPlaceholderText('Search');

    fireEvent.change(inputElement, { target: { value: 'new Value' } });

    expect(propsDefault.onChange).toBeCalledTimes(1);
  });

  it('should render match to snapshot default', () => {
    const { container } = render(<Search {...propsDefault} />);
    expect(container).toMatchSnapshot();
  });

  it('should render match to snapshot with icon', () => {
    const { container } = render(<Search {...propsDefault} icon={faSearch} />);
    expect(container).toMatchSnapshot();
  });
});
