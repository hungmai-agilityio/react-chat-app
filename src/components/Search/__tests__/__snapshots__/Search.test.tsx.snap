// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should render match to snapshot default 1`] = `
<div>
  <div
    class="relative"
  >
    <input
      class="border border-gray-300 focus:border-primary outline-none w-full h-10 pl-4 pr-9"
      placeholder="Search"
      value="value"
    />
  </div>
</div>
`;

exports[`Search should render match to snapshot with icon 1`] = `
<div>
  <div
    class="relative"
  >
    <input
      class="border border-gray-300 focus:border-primary outline-none w-full h-10 pl-4 pr-9"
      placeholder="Search"
      value="value"
    />
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-magnifying-glass absolute top-3 right-3 text-gray-500 cursor-pointer"
      data-icon="magnifying-glass"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z"
        fill="currentColor"
      />
    </svg>
  </div>
</div>
`;
